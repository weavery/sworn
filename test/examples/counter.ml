(* This is free and unencumbered software released into the public domain. *)

let fprintf = Format.fprintf

let read_file path =
  let channel = open_in path in
  let contents = really_input_string channel (in_channel_length channel) in
  close_in channel;
  contents

let () =
  let output_formatter = Format.formatter_of_out_channel stdout in
  let input = read_file "counter.clar" in
  let program = Clarity.parse_program input in
  let program' = SWIR.compile_program program in
  fprintf output_formatter "// Generated by Sworn@\n@\n";
  fprintf output_formatter "clarity.requireVersion(\"%s\")@\n@\n" "0.0";
  fprintf output_formatter "@[<v>%a@]@?" JavaScript.print_program program'
