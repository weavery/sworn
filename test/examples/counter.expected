// Generated by Sworn

clarity.requireVersion("0.0")

function getCounter(state) {
  return clarity.ok(state.counter);
}

function increment(state) {
  state.counter = clarity.add(state.counter, 1);
  return {state, result: clarity.ok(state.counter)};
}

function decrement(state) {
  state.counter = clarity.sub(state.counter, 1);
  return {state, result: clarity.ok(state.counter)};
}

export function handle(state, action) {
  const input = action.input;
  if (input.function === 'getCounter') {
    return {result: getCounter(state)};
  }
  if (input.function === 'increment') {
    return increment(state);
  }
  if (input.function === 'decrement') {
    return decrement(state);
  }
  return {state};
}
